src/VecSim/algorithms/hnsw/hnsw_batch_iterator.cpp

```diff
@@ -0,0 +1,206 @@
+#include "hnsw_batch_iterator.h"
+
+#include <utility>
+#include "VecSim/query_result_struct.h"
+#include "VecSim/utils/vec_utils.h"
+#include "VecSim/algorithms/hnsw/visited_nodes_handler.h"
+
+inline void HNSW_BatchIterator::visitNode(idType node_id) {
+    this->visited_list->tagNode(node_id, this->visited_tag);
+}
+
+inline bool HNSW_BatchIterator::hasVisitedNode(idType node_id) const {
+    return this->visited_list->getNodeTag(node_id) == this->visited_tag;
+}
+
+void HNSW_BatchIterator::prepareResults(
+    VecSimQueryResult_List batch_results,
+    vecsim_stl::max_priority_queue<pair<float, idType>> top_candidates, size_t n_res,
+    VecSimQueryResult_Order order) {
+    size_t initial_results_num = array_len(batch_results);
+    // Put the "spare" results (if exist) in the results heap.
```

Comment:
Under what conditions do we save results from the previous batch?

Reply:
If `ef_runtime` is larger than `n_res`, the graph scan returns the "spare" results.